elements/lisk-p2p/src/peer_directory/base_list.ts

```diff
@@ -0,0 +1,213 @@
+		return getBucket({
+			secret: this.peerListConfig.secret,
+			peerType: this.peerListConfig.peerType,
+			targetAddress: ipAddress,
+		});
+	}
+
+	public updatePeer(peerInfo: P2PPeerInfo): boolean {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this.peerMap.get(bucketId);
+
+		if (!bucket) {
+			return false;
+		}
+		const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+		const foundPeer = bucket.get(incomingPeerId);
+
+		if (!foundPeer) {
+			return false;
+		}
+		const updatedPeerInfo: CustomPeerInfo = {
+			...foundPeer,
+			peerInfo: { ...foundPeer.peerInfo, ...peerInfo },
+		};
+
+		bucket.set(incomingPeerId, updatedPeerInfo);
+		this.peerMap.set(bucketId, bucket);
+
+		return true;
+	}
+
+	public removePeer(peerInfo: P2PPeerInfo): boolean {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this.peerMap.get(bucketId);
+		const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+
+		if (bucket && bucket.get(incomingPeerId)) {
+			const success = bucket.delete(incomingPeerId);
+			this.peerMap.set(bucketId, bucket);
+
+			return success;
+		}
+
+		return false;
+	}
+
+	public getPeer(peerInfo: P2PPeerInfo): P2PPeerInfo | undefined {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this.peerMap.get(bucketId);
+		const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+
+		if (!bucket) {
+			return undefined;
+		}
+		const peer = bucket.get(incomingPeerId);
+
+		return peer ? peer.peerInfo : undefined;
+	}
+
+	public initPeerInfo = (peerInfo: P2PPeerInfo): CustomPeerInfo => ({
+		peerInfo,
+		dateAdded: new Date(),
+	});
+
+	// Addition of peer can also result in peer eviction if the bucket of the incoming peer is already full based on evection strategy.
+	public addPeer(peerInfo: P2PPeerInfo): AddPeerOutcome {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this.peerMap.get(bucketId);
+		const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+
+		if (!bucket) {
+			return {
+				success: false,
+				isEvicted: false,
+			};
+		}
+
+		if (bucket && bucket.get(incomingPeerId)) {
+			return {
+				success: false,
+				isEvicted: false,
+			};
+		}
+
+		const newPeer = this.initPeerInfo(peerInfo);
+
+		if (bucket.size < this.peerListConfig.peerBucketSize) {
+			bucket.set(incomingPeerId, newPeer);
+			this.peerMap.set(bucketId, bucket);
+
+			return {
+				success: true,
+				isEvicted: false,
+			};
+		}
+
+		const evictedPeer = this.evictPeer(bucketId);
+		bucket.set(incomingPeerId, newPeer);
+		this.peerMap.set(bucketId, bucket);
+
+		return {
+			success: true,
+			isEvicted: true,
+			evictedPeer: evictedPeer.peerInfo,
+		};
+	}
+
+	// This action is called when a peer is disconnected
+	public failedConnectionAction(incomingPeerInfo: P2PPeerInfo): boolean {
+		const success = this.removePeer(incomingPeerInfo);
+
+		return success;
+	}
+
+	public evictPeer(bucketId: number): CustomPeerInfo {
```

Comment:
Is this necessary if it's only wrapping?

Reply:
It seems like a wrapper in `base_list` but I'm using it as a function that will be used to override this function with your own eviction strategies. Whereas the function that is being called inside is just one eviction strategy that is by default in `base_list`

Reply:
I see. Can `evictRandomlyFromBucket` be protected in this case?

Reply:
Yes thats a good point, we can make it protected